### ISSUE TITLE
Support for compiling in Chinese paths and Chinese output directory, Chinese executes and Chinese module names not perfect yet

### ISSUE THREAD

**Issue by LINKlang** [Nuitka/Nuitka#2838]
- Nuitka version, full Python version
```
> python -m nuitka --version
2.2.1
Commercial: None
Python: 3.11.1 (tags/v3.11.1:a7a450f, Dec  6 2022, 19:58:39) [MSC v.1934 64 bit (AMD64)]
Flavor: CPython Official
Executable: C:\Users\ASUS\AppData\Local\Programs\Python\Python311\python.exe
OS: Windows
Arch: x86_64
WindowsRelease: 11
Nuitka-Scons: Non downloaded winlibs-gcc 'D:\llvm-mingw-20240502-ucrt-x86_64\bin\gcc.exe' is being ignored, Nuitka is
Nuitka-Scons: very dependent on the precise one.
Version C compiler: ~\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1\mingw64\bin\gcc.exe (gcc 13.2.0).
```
- How did you install Nuitka and Python
I installed Nuitka using pip.

Log
```
> nuitka --include-package-data=selenium --enable-plugin=tk-inter --standalone --onefile --windows-icon-from-ico=main.ico --output-dir="D:\Silly\CL - P\dist" "D:\Silly\CL - P\Cl.py"

Nuitka-Options: Used command line options: --include-package-data=selenium --enable-plugin=tk-inter --standalone
Nuitka-Options: --onefile --windows-icon-from-ico=main.ico --output-dir="D:\Silly\CL - P\dist" "D:\Silly\CL -
Nuitka-Options: P\Cl.py"
Nuitka: Starting Python compilation with Nuitka '2.2.1' on Python '3.11' commercial grade 'not installed'.
Nuitka-Plugins:WARNING: options-nanny: Using module 'attr' (version 0.3.2) with incomplete support due to condition
Nuitka-Plugins:WARNING: 'version("attrs") < (23,1)': 'attrs' is not fully supported before version 23.1.0
Nuitka-Plugins:options-nanny: Using module 'trio._core._ki' (version 0.22.0) with incomplete support: Disabled careful
Nuitka-Plugins:options-nanny: handling of KeyboardInterrupt in 'trio'
Nuitka-Plugins:options-nanny: Note, when using 'tkinter', consider using '--disable-console' option. Otherwise a
Nuitka-Plugins:options-nanny: terminal window will open. However for debugging, terminal output is the easiest way to
Nuitka-Plugins:options-nanny: see informative traceback and error information, so delay this until your program is
Nuitka-Plugins:options-nanny: working and remove once you find it non-working, and use '--enable-console' to make it
Nuitka-Plugins:options-nanny: explicit and not see this message.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka-Scons: Non downloaded winlibs-gcc 'D:\llvm-mingw-20240502-ucrt-x86_64\bin\gcc.exe' is being ignored, Nuitka is
Nuitka-Scons: very dependent on the precise one.
Nuitka will use gcc from MinGW64 of winlibs to compile on Windows.

Is it OK to download and put it in 'C:\Users\ASUS\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1'.

Fully automatic, cached. Proceed and download? [Yes]/No :
Nuitka: Downloading
Nuitka: 'https://github.com/brechtsanders/winlibs_mingw/releases/download/13.2.0-16.0.6-11.0.1-msvcrt-r1/winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-16.0.6-mingw-w64msvcrt-11.0.1-r1.zip'.
Nuitka: Extracting to
Nuitka: 'C:\Users\ASUS\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1\mingw64\bin\gcc.exe'
Nuitka-Scons: Backend C compiler: gcc (gcc 13.2.0).
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from __constants.c:2:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [__constants.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.OpenSSL.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.OpenSSL.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.OpenSSL._util.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from __helpers.c:4:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.OpenSSL._util.o] Error 1
scons: *** [__helpers.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.OpenSSL.SSL.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.OpenSSL.SSL.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.OpenSSL.crypto.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.OpenSSL.version.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.OpenSSL.crypto.o] Error 1
scons: *** [module.OpenSSL.version.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.async_generator.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.async_generator.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.attr._compat.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.attr._compat.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.async_generator._version.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.async_generator._version.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.__parents_main__.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.async_generator._impl.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.async_generator._util.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.__parents_main__.o] Error 1
scons: *** [module.async_generator._impl.o] Error 1
scons: *** [module.async_generator._util.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.attr._cmp.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.attr._cmp.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.__main__.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.__main__.o] Error 1
In file included from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/exceptions.h:10,
                 from C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/prelude.h:431,
                 from module.attr._config.c:19:
C:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\Lib\SITE-P~1\nuitka\build\include/nuitka/constants.h:216:10: fatal error: __constants.h: No such file or directory
  216 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.

scons: *** [module.attr._config.o] Error 1
FATAL: Failed unexpectedly in Scons C backend compilation.
Nuitka:WARNING:     Complex topic! More information can be found at
Nuitka:WARNING: https://nuitka.net/info/scons-backend-failure.html
Nuitka-Reports: Compilation crash report written to file 'nuitka-crash-report.xml'.
```

**Comment by kayhayen** [comment]
I wonder if the rejected gcc is causing this, or if the output directory with spaces is causing it, it seems the source files are found, and the constants file would be generated in the same spot.

**Comment by kayhayen** [comment]
Can you provide the output when compiling with `--show-scons` option please?

**Comment by LINKlang** [comment]
I think I understand the reason for this issue now. It seems to be because the path contains Chinese characters.

> 我想知道是被拒绝的gcc导致了这种情况，还是带有空格的输出目录导致了这种情况，似乎找到了源文件，并且常量文件将在同一位置生成。

**Comment by LINKlang** [comment]
> I wonder if the rejected gcc is causing this, or if the output directory with spaces is causing it, it seems the source files are found, and the constants file would be generated in the same spot.

I think I understand the reason for this issue now. It seems to be because the path contains Chinese characters.

**Comment by kayhayen** [comment]
Which you carefully removed from your report, right? I would be interested where they show up, because they are supposed not to, and --show-scons will help me see it.

**Comment by LINKlang** [comment]
[log.txt](https://github.com/Nuitka/Nuitka/files/15272191/log.txt)


> Which you carefully removed from your report, right? I would be interested where they show up, because they are supposed not to, and --show-scons will help me see it.

**Comment by kayhayen** [comment]
Thanks, very helpful. It's actually that the --output-dir is chinese, so `result_name` and `result_exe` get passed with Chinese characters to scons. That happens in both the path and the base filename. I am not sure it would resolve that properly, but likely it will, but that's only at link time.

The file the C compiler does not find however, strangely lives in `.` and I don't see how that could go wrong. I will have to try and use --output-dir with that kind of path, maybe I can then reproduce it.

**Comment by LINKlang** [comment]
> Thanks, very helpful. It's actually that the --output-dir is chinese, so `result_name` and `result_exe` get passed with Chinese characters to scons. That happens in both the path and the base filename. I am not sure it would resolve that properly, but likely it will, but that's only at link time.
> 
> The file the C compiler does not find however, strangely lives in `.` and I don't see how that could go wrong. I will have to try and use --output-dir with that kind of path, maybe I can then reproduce it.

If there's anything I can assist you with, please let me know.

**Comment by kayhayen** [comment]
I managed to reproduce an issue for ccache with this, compiling in a chinese directory, is translated well, but the output dir is not, it seems: ```ccache: error: Failed to write to C:\Users\kayha\repos\CCA1E~1\???C\dist\AssertsTest.build\ccache-34228.txt: Invalid argument```, so that's an angle to resolve things.

**Comment by kayhayen** [comment]
So, the issue is that `.` is expanded in scons, to the full chinese path. For me, `result_exe` and `result_name` are not chinese, but correct, not sure why they are for you, maybe the final program filename is it. That would indeed possibly pose a challenge hard to overcome.

**Comment by kayhayen** [comment]
Using a chinese filename, I am getting this from MinGW

> __loader.c:40:1: warning: hex escape sequence out of range

It seems a value of `344828` is used there, according to MSVC. 

That seems to only happen in accelerated mode, where the original paths are stored for use inside the binary, and it reveals a mistake in how unicode strings are encoding for use in C there.

**Comment by kayhayen** [comment]
So, next up the CMD script couldn't handle unicode, but that can be worked around. With that MSVC is working fine. But MinGW64 will not handle `-o C:\Users\kayha\repos\CCA1E~1\CCA1E~1\dist\开启.exe` and will need some help to do that it seems, like we supply a short path, and pre-create the file, so the filenames alias, but I am not sure that can work.

**Comment by kayhayen** [comment]
So, I am adding workarounds, basically enforcing the use of short names for MinGW64 but we need to create an empty file to know the shortname, and then it seems we can make MinGW64 overwrite it. With that, executables work with Chinese result names, modules I also changed, but for some reason, we end up having C++ detected as the language or our C files, and this gives an error. No idea why yet, but it could well fail to detect things inside of unicode paths there.

**Comment by kayhayen** [comment]
So, cleaning up things, Nuitka in Scons now checks if the target filename is encodable, and if not, changes it to "_nuitka_temp.pyd" or "_nuitka_temp.exe" and after successfully compiling, uses a rename operation to go to the real needed name.

Then, it seems to be a similar problem, where dependency walker was using the basename of the created binary, for temp files it uses, so we do something similar there, these names are only there to easily associate things for debugging, so onefile works now in this setup too.

One issue remaining, is that MinGW64 doesn't recognize files as C from their name, e.g. `C:\Users\kayha\repos\CCA1E~1\CCA1E~1\dist\4ADC~1.BUI\MODULE~1.C` when I use a Chine output directory, I am researching that now.

**Comment by kayhayen** [comment]
So, compiling now works for modules even with output directories, but now I get this:

```
Nuitka: Launching '键开启C\dist\开启.cp310-win_amd64.pyd'.
Traceback (most recent call last):
  File "<string>", line 2, in <module>
ImportError: dynamic module does not define module export function (PyInitU_drrp5w)
```

The name of the init function and the module name have to match. It seems that `U_drrp5w` is some sort of encoding that Python uses, most likely DLL entry points, cannot be unicode, and that's their way of working around that, but Nuitka doesn't yet know that kind of code.

**Comment by kayhayen** [comment]
So, it seems it uses something called `punycode` for the entry points, and now it was complaining about it not returning a 2 phase init object, and not having per sub-interpreter support for new style loads. So I added that as well, and now it works for extension modules. The created extension modules now load fine into the interpreter, but I expect that Nuitka standalone won't be able to make use of them until we mirror the entry point name change there too.

**Comment by kayhayen** [comment]
So, I added that entry point change, and the 2 phase module load successfully as well. And then there was also an issue with the filename of DLL being wrong, which has been corrected as well. At some point, I also saw the original error you had, since it's using `.' as source_dir in some configurations, and otherwise the source_dir not shortened, this can happen as well.

I am putting this through my internal CI now, and then we should have something which is a lot closer to working for many more Chinese setups. I am still a bit confused why `result_exe` had non-short paths in there. I am suspecting a bug in the caching of short paths, that I want to look at.

**Comment by kayhayen** [comment]
Thanks for your report, this is fixed on the factory branch, which is a development version under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 

Feedback on whether this is working is very welcome. Please do not share plans to do it; only confirm or deny that it is working.

**Comment by kayhayen** [comment]
It would seem, that you should be able to compile in your environment with the exact options successfully now with the factory version. Let me know if that's true. When something fails, I would love to have a report. I found huge amounts of fixes, surprising, mostly related to unicode output filenames, extension modules, etc. and I would love to see if things improved for you. I care much about out of the box experience for Nuitka on Asian systems.

**Comment by susundingkai** [comment]
Not working under Macos too, you can reproduce the problem like this:
`with open("test.txt","a+") as fp: fp.write("测试")`

**Comment by kayhayen** [comment]
That doesn't look like Python code, but a data file, when I renamed it to `.py` which is what you intended probably, I indeed got a crash from the `NameError`. So that's more like core Nuitka, checking it out.

**Comment by kayhayen** [comment]
So, I have updated factory with fixes for this. The code was assuming it can always take a UTF8 string for the exception, but that's not true, and for Python3 there is actually a better method to use, so I applied this everywhere I saw similar code, function repr, class names, method repr, local variable errors, etc.

This actually could be part of a hotfix, but I feel that 2.3 is approaching fast, and I had to make an emergency one for PySide6 today, so I don't think there will be another one necessarily, and then it's not work the risk. Right now my CI has better assertions on, and you can try `--debug` if you still have crashes, should catch those cases better.

**Comment by susundingkai** [comment]
Strangely, it work well in debug mode and I can not get any crash log.

**Comment by susundingkai** [comment]
I found that I can execute test.app/Contents/main, but cant run test.app directly.  Maybe the console-less mode caused breaking down?

**Comment by kayhayen** [comment]
Actually `--disable-console´ has had no effect on macOS for a while now. What do you mean with "directly"? How sure are we that macOS itself handles this fine, or that signing with Chinese filenames works. But it seems you are hijacking this issue, which was about Windows. Please make a new issue @susundingkai esp one with a reproducer.

**Comment by susundingkai** [comment]
> Actually `--disable-console´ has had no effect on macOS for a while now. What do you mean with "directly"? How sure are we that macOS itself handles this fine, or that signing with Chinese filenames works. But it seems you are hijacking this issue, which was about Windows. Please make a new issue @susundingkai esp one with a reproducer.

I mean, open the test.app folder, find the executable file, and double click it to execute the file. It will open a console and run without crash.

**Comment by kayhayen** [comment]
The variable names related stuff is now on develop and will be in 2.3 release as well, which is bound to happen any day now.

**Comment by LINKlang** [comment]
> Thanks for your report, this is fixed on the factory branch, which is a development version under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html
> 
> Feedback on whether this is working is very welcome. Please do not share plans to do it; only confirm or deny that it is working.

I tried installing the factory branch using `pip install -e git+https://github.com/Nuitka/Nuitka.git@factory#egg=nuitka`, but after installation, Nuitka still couldn't find `__constants.h`.

**Comment by kayhayen** [comment]
That's too bad, can you provide the output with --show-scons again?

**Comment by LINKlang** [comment]
> That's too bad, can you provide the output with --show-scons again?

[log.txt](https://github.com/Nuitka/Nuitka/files/15496596/log.txt)

**Comment by kayhayen** [comment]
Ok, I think it's recognizable, that the translation to short paths somehow fails there. Both `result_exe` and `nuitka_src` are being wrong. The call of the scons file and the Python for the scons itself, might also be wrong. Maybe Nuitka does caching that is broken or maybe the shortname translation fails for your specific paths. I am seeing that there is this code:

```
            # Permission denied.
            if ctypes.GetLastError() == 5:
                return filename
```

Can you comment that out in `nuitka.utils.FileOperation` and see if you instead get a Python exception? Not sure, why permission could be denied there, but maybe short paths are not really always available to use and would require priviledge.

**Comment by LINKlang** [comment]
> Ok, I think it's recognizable, that the translation to short paths somehow fails there. Both `result_exe` and `nuitka_src` are being wrong. The call of the scons file and the Python for the scons itself, might also be wrong. Maybe Nuitka does caching that is broken or maybe the shortname translation fails for your specific paths. I am seeing that there is this code:
> 
> ```
>             # Permission denied.
>             if ctypes.GetLastError() == 5:
>                 return filename
> ```
> 
> Can you comment that out in `nuitka.utils.FileOperation` and see if you instead get a Python exception? Not sure, why permission could be denied there, but maybe short paths are not really always available to use and would require priviledge.

![image](https://github.com/Nuitka/Nuitka/assets/116951234/7ac07292-12cc-4aaa-98da-4965cc24e85c)
I tried commenting out the code, but the same error occurred.
[log.txt](https://github.com/Nuitka/Nuitka/files/15497705/log.txt)

**Comment by kayhayen** [comment]
I don't really know what could be causing this. My current fixes are part of 2.3 release, but it seems there is something to get to the bottom of still. 

There is a line `options["nuitka_src"] = getExternalUsePath(options["nuitka_src"])` and that seems to either not run for you, but I am seeing the impact of `options["source_dir"] = "."` so I think it is, and somehow said function, which you commented out code on its path for me now, is not running, or we are being dumb. But the version information seems to imply you have the same code as I there. Use 2.3 now and let me know, if you manage to somehow figure out how it gets to return an absolute path with Chinese characters there.

**Comment by LINKlang** [comment]
> I don't really know what could be causing this. My current fixes are part of 2.3 release, but it seems there is something to get to the bottom of still.
> 
> There is a line `options["nuitka_src"] = getExternalUsePath(options["nuitka_src"])` and that seems to either not run for you, but I am seeing the impact of `options["source_dir"] = "."` so I think it is, and somehow said function, which you commented out code on its path for me now, is not running, or we are being dumb. But the version information seems to imply you have the same code as I there. Use 2.3 now and let me know, if you manage to somehow figure out how it gets to return an absolute path with Chinese characters there.

![image](https://github.com/Nuitka/Nuitka/assets/116951234/29051e2d-1dec-4cc2-bafa-b04af64f522e)
It outputted:

```
Before: D:\Silly\真一键脚本 - 抢先版\venv\src\nuitka\nuitka\build
After: D:\Silly\真一键~3\venv\src\nuitka\nuitka\build
```

I tried using `D:\Silly\真一键~3`, which is the correct short path for Chinese characters.

**Comment by kayhayen** [comment]
Ah, it is? But I have serious doubts that encoding-wise, MinGW64 does well there. Short Chinese paths are a huge surprise to me. I thought the whole idea was not for them to be short but for them to be ASCII. I need to study the MS docs there.

**Comment by kayhayen** [comment]
So, it seems nothing in the docs guarantees this to be anything but a shorter path, lengthwise. However, in my testing, e.g., Chinese extension module names became usable through this. Here, the current directory is where the included file lives, and we switch to the short path. Then GCC reads it and seemingly cannot handle it, with it being Chinese still. So, where we use "." there, not even relative paths are working out.

What's striking, though, is that this is the only relative path used, and that should cause it to not be found. That's very strange, is it not?

In this command line:

```
"C:\Users\ASUS\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1\mingw64\bin\ccache.exe" "C:\Users\ASUS\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1\mingw64\bin\gcc.exe" -o module.attr.setters.o -c -std=c11 -fvisibility=hidden -fwrapv -pipe -fpartial-inlining -ftrack-macro-expansion=0 -Wno-deprecated-declarations -fno-var-tracking -Wno-misleading-indentation -fcompare-debug-second -O3 -D_WIN32_WINNT=0x0501 -D__NUITKA_NO_ASSERT__ -DMS_WIN64 -D_NUITKA_STANDALONE -D_NUITKA_ONEFILE_MODE -D_NUITKA_ONEFILE_TEMP_BOOL -D_NUITKA_CONSTANTS_FROM_RESOURCE -D_NUITKA_FROZEN=153 -D_NUITKA_EXE -D_NUITKA_PLUGIN_MULTIPROCESSING_ENABLED=1 -ID:\Silly\真一键~3\venv\src\nuitka\nuitka\build\inline_copy\zlib -IC:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\include -I. -ID:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include -
```

The part not working there would be the `-I.` part. The absolute paths to parts of Nuitka did come into play already, it's loading `__constants.h` from that.

See this

```
D:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include/nuitka/constants.h:220:10: fatal error: __constants.h: No such file or directory
  220 | #include "__constants.h"
```

So either this file really doesn't exist in the build folder, please confirm, or the fact that it is `.` which should be a short path too, somehow gets lost, and then that's an issue for MinGW64. I swear I saw MinGW64 fail when Nuitka did live in a Chinese path, and making it short helped.

To verify, you could change this:

```
            options["source_dir"] = "."
            options["nuitka_src"] = getExternalUsePath(options["nuitka_src"])
```

to 

```
            options["source_dir"] = getExternalUsePath(options["source_dir"])
            options["nuitka_src"] = getExternalUsePath(options["nuitka_src"])
```

That should change the `-I.` and make it an absolute path, too. These are actively avoided to achieve caching across different build folders, but it might be what triggers the issue for you. The Chinese short paths seem to work. Maybe MinGW64 creates a long form path out of `.` or the Windows process forked this way:

```
            if source_dir is not None:
                _createSconsDebugScript(
                    source_dir=source_dir, scons_command=scons_command
                )

                source_dir = getExternalUsePath(source_dir)
            try:
                result = subprocess.call(scons_command, shell=False, cwd=source_dir)
            except KeyboardInterrupt:
                Tracing.scons_logger.sysexit("User interrupted scons build.")
```

So, please if you can try these things, I think I am very close to understanding this one.

**Comment by LINKlang** [comment]
> So, it seems nothing in the docs guarantees this to be anything but a shorter path, lengthwise. However, in my testing, e.g., Chinese extension module names became usable through this. Here, the current directory is where the included file lives, and we switch to the short path. Then GCC reads it and seemingly cannot handle it, with it being Chinese still. So, where we use "." there, not even relative paths are working out.
> 
> What's striking, though, is that this is the only relative path used, and that should cause it to not be found. That's very strange, is it not?
> 
> In this command line:
> 
> ```
> "C:\Users\ASUS\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1\mingw64\bin\ccache.exe" "C:\Users\ASUS\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1\mingw64\bin\gcc.exe" -o module.attr.setters.o -c -std=c11 -fvisibility=hidden -fwrapv -pipe -fpartial-inlining -ftrack-macro-expansion=0 -Wno-deprecated-declarations -fno-var-tracking -Wno-misleading-indentation -fcompare-debug-second -O3 -D_WIN32_WINNT=0x0501 -D__NUITKA_NO_ASSERT__ -DMS_WIN64 -D_NUITKA_STANDALONE -D_NUITKA_ONEFILE_MODE -D_NUITKA_ONEFILE_TEMP_BOOL -D_NUITKA_CONSTANTS_FROM_RESOURCE -D_NUITKA_FROZEN=153 -D_NUITKA_EXE -D_NUITKA_PLUGIN_MULTIPROCESSING_ENABLED=1 -ID:\Silly\真一键~3\venv\src\nuitka\nuitka\build\inline_copy\zlib -IC:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\include -I. -ID:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include -
> ```
> 
> The part not working there would be the `-I.` part. The absolute paths to parts of Nuitka did come into play already, it's loading `__constants.h` from that.
> 
> See this
> 
> ```
> D:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include/nuitka/constants.h:220:10: fatal error: __constants.h: No such file or directory
>   220 | #include "__constants.h"
> ```
> 
> So either this file really doesn't exist in the build folder, please confirm, or the fact that it is `.` which should be a short path too, somehow gets lost, and then that's an issue for MinGW64. I swear I saw MinGW64 fail when Nuitka did live in a Chinese path, and making it short helped.
> 
> To verify, you could change this:
> 
> ```
>             options["source_dir"] = "."
>             options["nuitka_src"] = getExternalUsePath(options["nuitka_src"])
> ```
> 
> to
> 
> ```
>             options["source_dir"] = getExternalUsePath(options["source_dir"])
>             options["nuitka_src"] = getExternalUsePath(options["nuitka_src"])
> ```
> 
> That should change the `-I.` and make it an absolute path, too. These are actively avoided to achieve caching across different build folders, but it might be what triggers the issue for you. The Chinese short paths seem to work. Maybe MinGW64 creates a long form path out of `.` or the Windows process forked this way:
> 
> ```
>             if source_dir is not None:
>                 _createSconsDebugScript(
>                     source_dir=source_dir, scons_command=scons_command
>                 )
> 
>                 source_dir = getExternalUsePath(source_dir)
>             try:
>                 result = subprocess.call(scons_command, shell=False, cwd=source_dir)
>             except KeyboardInterrupt:
>                 Tracing.scons_logger.sysexit("User interrupted scons build.")
> ```
> 
> So, please if you can try these things, I think I am very close to understanding this one.

I think I may have discovered something. I checked the `D:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include/nuitka` path, but I did not see `__constants.h`. (It's also possible I don't understand C code? Is it normal for this file to be missing?)
![image](https://github.com/Nuitka/Nuitka/assets/116951234/596ebed7-1071-4fee-871b-f98d44a2cfab)

So next, I will try modifying `getExternalUsePath(options["source_dir"])`.

**Comment by kayhayen** [comment]
No, that would be normal, it's supposed to exist in the build folder, where it is generated into for the specific program. 

When I reproduced the other issues, it was not true, but because MinGW64 can not use non-Unicode paths for that build folder. It is a short path, but still Chinese, I think, is what is causing the symptoms. However, I also observed, that it is the only path that is not used absolute, and that short Chinese paths appear to be working. 

I don't understand why I didn't have short Chinese paths. Maybe because I have a non-Chinese locale, my NTFS doesn't do them at all. Maybe then work with a Chinese locale, but not as relative paths, or something like that. That is why I suggested the change to make, and also kind of expect it to help your case.

**Comment by LINKlang** [comment]
> So, it seems nothing in the docs guarantees this to be anything but a shorter path, lengthwise. However, in my testing, e.g., Chinese extension module names became usable through this. Here, the current directory is where the included file lives, and we switch to the short path. Then GCC reads it and seemingly cannot handle it, with it being Chinese still. So, where we use "." there, not even relative paths are working out.
> 
> What's striking, though, is that this is the only relative path used, and that should cause it to not be found. That's very strange, is it not?
> 
> In this command line:
> 
> ```
> "C:\Users\ASUS\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1\mingw64\bin\ccache.exe" "C:\Users\ASUS\AppData\Local\Nuitka\Nuitka\Cache\DOWNLO~1\gcc\x86_64\13.2.0-16.0.6-11.0.1-msvcrt-r1\mingw64\bin\gcc.exe" -o module.attr.setters.o -c -std=c11 -fvisibility=hidden -fwrapv -pipe -fpartial-inlining -ftrack-macro-expansion=0 -Wno-deprecated-declarations -fno-var-tracking -Wno-misleading-indentation -fcompare-debug-second -O3 -D_WIN32_WINNT=0x0501 -D__NUITKA_NO_ASSERT__ -DMS_WIN64 -D_NUITKA_STANDALONE -D_NUITKA_ONEFILE_MODE -D_NUITKA_ONEFILE_TEMP_BOOL -D_NUITKA_CONSTANTS_FROM_RESOURCE -D_NUITKA_FROZEN=153 -D_NUITKA_EXE -D_NUITKA_PLUGIN_MULTIPROCESSING_ENABLED=1 -ID:\Silly\真一键~3\venv\src\nuitka\nuitka\build\inline_copy\zlib -IC:\Users\ASUS\AppData\Local\Programs\Python\PYTHON~1\include -I. -ID:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include -
> ```
> 
> The part not working there would be the `-I.` part. The absolute paths to parts of Nuitka did come into play already, it's loading `__constants.h` from that.
> 
> See this
> 
> ```
> D:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include/nuitka/constants.h:220:10: fatal error: __constants.h: No such file or directory
>   220 | #include "__constants.h"
> ```
> 
> So either this file really doesn't exist in the build folder, please confirm, or the fact that it is `.` which should be a short path too, somehow gets lost, and then that's an issue for MinGW64. I swear I saw MinGW64 fail when Nuitka did live in a Chinese path, and making it short helped.
> 
> To verify, you could change this:
> 
> ```
>             options["source_dir"] = "."
>             options["nuitka_src"] = getExternalUsePath(options["nuitka_src"])
> ```
> 
> to
> 
> ```
>             options["source_dir"] = getExternalUsePath(options["source_dir"])
>             options["nuitka_src"] = getExternalUsePath(options["nuitka_src"])
> ```
> 
> That should change the `-I.` and make it an absolute path, too. These are actively avoided to achieve caching across different build folders, but it might be what triggers the issue for you. The Chinese short paths seem to work. Maybe MinGW64 creates a long form path out of `.` or the Windows process forked this way:
> 
> ```
>             if source_dir is not None:
>                 _createSconsDebugScript(
>                     source_dir=source_dir, scons_command=scons_command
>                 )
> 
>                 source_dir = getExternalUsePath(source_dir)
>             try:
>                 result = subprocess.call(scons_command, shell=False, cwd=source_dir)
>             except KeyboardInterrupt:
>                 Tracing.scons_logger.sysexit("User interrupted scons build.")
> ```
> 
> So, please if you can try these things, I think I am very close to understanding this one.

![image](https://github.com/Nuitka/Nuitka/assets/116951234/e0cf2caa-b4d7-4d52-81b4-40b5690bbfbf)

I have changed the code here, but the error still occurs. I noticed that the log indicates Nuitka is still using the short path, which might be the problem. Additionally, I found that after modifying the code, the `.build` and `.dist` folders were correctly generated in the dist directory, which wasn't happening before.

![image](https://github.com/Nuitka/Nuitka/assets/116951234/e5dc5638-ca11-4a5e-80b2-6470eb50a219)

**Comment by kayhayen** [comment]
It seems that the `__constants.h` and `__constants.c` files are there, yet the MinGW64 does not find them in an include path scan, although it does find the request to include nuitka_src from a Chinese short path. 

We lack an explanation of how it could be. My guess that `.` is not good for them, maybe expanding that to a long Chinese path it cannot handle, should be out the Window, if you used the external path for it, i.e. passed as absolute. I am out of ideas honestly.

**Comment by LINKlang** [comment]
[log.txt](https://github.com/user-attachments/files/15716863/log.txt)


> It seems that the `__constants.h` and `__constants.c` files are there, yet the MinGW64 does not find them in an include path scan, although it does find the request to include nuitka_src from a Chinese short path.
> 
> We lack an explanation of how it could be. My guess that `.` is not good for them, maybe expanding that to a long Chinese path it cannot handle, should be out the Window, if you used the external path for it, i.e. passed as absolute. I am out of ideas honestly.

I have uploaded the log, which might be helpful. I am now wondering why the compiler is still using the short path for lookup even after using `getExternalUsePath(options["source_dir"])`.

**Comment by kayhayen** [comment]
That is what it does, turning a path absolute and making it short. Long chinese paths are known to not work with MinGW64, and it seems short paths always work.

The gist of the issue is here:

```
-ID:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include -ID:\Silly\真一键~3\venv\src\nuitka\nuitka\build\static_src -ID:\Silly\真一键~3\venv\src\nuitka\nuitka\build\inline_copy\libbacktrace module.attr.setters.c
```

Notice, how it tells it to look in short paths, and how MinGW64 does so successfully:

```
In file included from D:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include/nuitka/exceptions.h:10,
                 from D:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include/nuitka/prelude.h:464,
                 from __helpers.c:4:
D:\Silly\真一键~3\venv\src\nuitka\nuitka\build\include/nuitka/constants.h:220:10: fatal error: __constants.h: No such file or directory
```

Then I fails to find this file:

```
  220 | #include "__constants.h"
      |          ^~~~~~~~~~~~~~~
compilation terminated.
```

And that is from `-I.` which it ought to not use I guess, because this is asking it to go from the current directory (which is a short chinese path, and it seems to fail at that. This is `source_dir` though and if you did the changes I suggest, it would not be "." but an external path. I think you maybe didn't do what I asked you to do this time? Make sure to not see -I. in the scons output.

Yours,
Kay

**Comment by kayhayen** [comment]
It's Python code, I cannot remote debug it. It could be that `shallCompileWithoutBuildDirectory` is not true for you, if you use --debugger, that would be the case.

**Comment by LINKlang** [comment]
> It's Python code, I cannot remote debug it. It could be that `shallCompileWithoutBuildDirectory` is not true for you, if you use --debugger, that would be the case.

To be honest, I think Nuitka has done its job; the issue seems to lie with MinGW64. Can you provide a compressed package of MinGW64 for me to check if the issue is with my MinGW64?

**Comment by kayhayen** [comment]
The URL for download is in the source code of Nuitka. If you are not using that exact MinGW64, then of course you may have bugs that I do not have. It being buggy is unfortunately extremely common, which is why Nuitka rejects non-downloaded gccs on Windows, if you cheated there, then of course it's all off and a waste of time. In case it's not, too bad, but I will have to wait for a report with a genuine install.

**Comment by LINKlang** [comment]
> The URL for download is in the source code of Nuitka. If you are not using that exact MinGW64, then of course you may have bugs that I do not have. It being buggy is unfortunately extremely common, which is why Nuitka rejects non-downloaded gccs on Windows, if you cheated there, then of course it's all off and a waste of time. In case it's not, too bad, but I will have to wait for a report with a genuine install.

If you mean the MingW64 from `https://github.com/brechtsanders/winlibs_mingw/releases/download/13.2.0-16.0.6-11.0.1-msvcrt-r1/winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-16.0.6-mingw-w64msvcrt-11.0.1-r1.zip`, I've already tried it. 😥
[log.txt](https://github.com/user-attachments/files/15882501/log.txt)

**Comment by kayhayen** [comment]
You have not done what's needed to not have -I in the scons calls of the gcc. My theory is that it will be better, but it seems you do not manage. That's OK, lets wait until somebody with the same issue can.

**Comment by LINKlang** [comment]
> You have not done what's needed to not have -I in the scons calls of the gcc. My theory is that it will be better, but it seems you do not manage. That's OK, lets wait until somebody with the same issue can.

I upgraded Nuitka to version 2.3.11 and changed MinGW to use `13.2.0-16.0.6-11.0.1-msvcrt-r1`. If I'm still doing something wrong, it seems I might not be suited to using Nuitka.

**Comment by kayhayen** [comment]
You don't have to change anything about MinGW, but let it download. There will be another chance once I switch Nuitka to an Anaconda-based setup in the 2.5 time frame, and maybe their gcc, but yeah, let's give up for now. Even though your report is great, your debugging ability is not there, and my remote debugging skills are equally insufficient.
